src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -120,171 +110,176 @@ public void Dispose()
         {
             CheckIfDisposed();
 
-            lock (_guard)
+            CompilerPathUtilities.RequireAbsolutePath(originalPath, nameof(originalPath));
+            var simpleName = PathUtilities.GetFileName(originalPath, includeExtension: false);
+            string realPath = originalPath;
+            IAnalyzerPathResolver? resolver = null;
+            foreach (var current in AnalyzerPathResolvers)
             {
-                return _analyzerAssemblyInfoMap.ContainsKey(fullPath);
+                if (current.IsAnalyzerPathHandled(originalPath))
+                {
+                    resolver = current;
+                    realPath = resolver.GetRealAnalyzerPath(originalPath);
+                    break;
+                }
             }
-        }
-
-        public void AddDependencyLocation(string fullPath)
-        {
-            CheckIfDisposed();
 
-            CompilerPathUtilities.RequireAbsolutePath(fullPath, nameof(fullPath));
-            string simpleName = PathUtilities.GetFileName(fullPath, includeExtension: false);
+            var assemblyName = readAssemblyName(realPath);
 
             lock (_guard)
             {
-                if (!_knownAssemblyPathsBySimpleName.TryGetValue(simpleName, out var paths))
+                if (_originalPathInfoMap.TryAdd(originalPath, (resolver, realPath, assemblyName)))
+                {
+                    // In the case multiple original paths map to the same real path then the first on
+                    // wins.
+                    //
+                    // An example reason to map multiple original paths to the same real path would be to
+                    // unify references.
+                    _ = _realToOriginalPathMap.TryAdd(realPath, originalPath);
+
+                    if (!_assemblySimpleNameToOriginalPathListMap.TryGetValue(simpleName, out var set))
+                    {
+                        set = new();
+                        _assemblySimpleNameToOriginalPathListMap[simpleName] = set;
+                    }
+
+                    _ = set.Add(originalPath);
+                }
+            }
+
+            static AssemblyName? readAssemblyName(string filePath)
+            {
+                AssemblyName? assemblyName;
+                try
                 {
-                    paths = ImmutableHashSet.Create(PathUtilities.Comparer, fullPath);
-                    _knownAssemblyPathsBySimpleName.Add(simpleName, paths);
+                    assemblyName = AssemblyName.GetAssemblyName(filePath);
                 }
-                else
+                catch
                 {
-                    _knownAssemblyPathsBySimpleName[simpleName] = paths.Add(fullPath);
+                    // The above can fail when the assembly doesn't exist because it's corrupted, 
+                    // doesn't exist on disk, or is a native DLL. Those failures are handled when 
+                    // the actual load is attempted. Just record the failure now.
```

Comment:
fwiw, i would prefer distinct exception catches for each kind of exception (instead of a blanket catch all).  the latter ends up catching product bugs too often.  in this case, as a single liner that direclty calls into the runtime, i guess it's ok.  but in general i've been burned :)

Reply:
I've been burned the other way: the set of thrown exceptions being incorrectly documented or simply changing release to release. For cases like this I ask myself

> Is there any exception type that could be thrown here that would cause be to take a different action?

If the answer is "no" then `catch` all cause otherwise it's servicing fixes to add another exception to the `catch` clause. This is a case where it's not really interesting why it failed, just that it failed.